salt/iiif/init.sls

```diff
@@ -0,0 +1,224 @@
+        - group: {{ loris_user }}
+        - dir_mode: 755
+        - makedirs: True
+        - require:
+            - loris-user
+            - mount-external-volume
+
+
+# Docker needs to bind certain configuration files between host and container
+
+
+loris-dir:
+    file.directory:
+        - user: {{ pillar.elife.deploy_user.username }}
+        - name: /opt/loris
+
+loris-config:
+    file.managed:
+        - user: {{ pillar.elife.deploy_user.username }}
+        - name: /opt/loris/loris2.conf
+        - source: salt://iiif/config/opt-loris-loris2.conf
+        - template: jinja
+        - makedirs: True
+        - require:
+            - loris-dir
+
+# required by newrelic-python.sls as it uses the 'newrelic' installed in the venv to generate the licence file
+loris-newrelic-venv:
+    cmd.run:
+        - cwd: /opt/loris
+        - runas: {{ pillar.elife.deploy_user.username }}
+        - name: |
+            python3 -m venv venv
+            venv/bin/pip install newrelic==5.8.0.136
+        - unless:
+            - test -d /opt/loris/venv
+            
+# required by newrelic-python.sls because it's using builder-private and not the formula's pillar
+# todo: remove once builder-private changes are in and the service is removed
+loris-uwsgi-ready:
+    service.running:
+        - name: nginx # could be anything that should be enabled by default
+
+# required by newrelic-python.sls because it's using builder-private and not the formula's pillar
+# todo: remove once builder-private changes are in
+loris-setup:
+    cmd.run:
+        - name: "echo dummy state"
+        - require:
+            - loris-config
+            - loris-newrelic-venv
+            - loris-uwsgi-ready
+
+
+# newrelic-python.sls starts running about here
+
+
+loris-uwsgi-config:
+    file.managed:
+        - name: /opt/loris/uwsgi.ini
+        - source: salt://iiif/config/opt-loris-uwsgi.ini
+        - template: jinja
+        - require:
+            - loris-dir
+
+loris-wsgi-entry-point:
+    file.managed:
+        - name: /opt/loris/loris2.wsgi
+        - source: salt://iiif/config/opt-loris-loris2.wsgi
+        - require:
+            - loris-dir
+
+
+
+{% if pillar.elife.env == "dev" %}
+
+# good for development.
+# just clone or move the 'loris-docker' repository into the root of your builder installation.
+build-loris:
+    docker_image.present:
+        - name: elifesciences/loris
+        - tag: latest
+        - build: /vagrant/loris-docker
+        - force: true
+        - require_in:
+            - docker_container: run-loris
+        - onlyif:
+            - test -d /vagrant/loris-docker
+
+{% endif %}        
+
+# `docker_container.running` state:
+# - https://docs.saltstack.com/en/latest/ref/states/all/salt.states.docker_container.html#salt.states.docker_container.running
+run-loris:
+    docker_container.running:
+        - name: loris--{{ pillar.elife.env }} # loris--dev, loris--prod
+        - image: elifesciences/loris
```

Comment:
`:latest` is implied?

Reply:
it is implied, but apparently other magic happens if it's not explicit:
https://docs.saltstack.com/en/latest/ref/states/all/salt.states.docker_container.html#salt.states.docker_container.running

I'll make the tag explicit